### DISCUSSION TITLE
What's the planned scope of Widgets for owlkettle?

### DISCUSSION THREAD

**Discussion by PhilippMDoerner** [can-lehmann/owlkettle#100] [Development]
During the wrapping of widgets like StatusPage and Expander I noticed that they were pretty easy and seemed like pretty "low-effort, high return" widgets to include. Particularly compared to stuff like SpinRow or Video where one I'm starting to think is just beyond my understanding and the other also requires wrapping the MediaStream types.

So that got me wondering for the time after at least the Video Widget is done: Given that there's a decent chunk of Widgets of GTK/Adwaita not yet in Owlkettle, is it worth looking for some more low-hanging-fruit like those and adding them even without an issue present?

@can-lehmann what would be your thoughts on that?

**Comment by can-lehmann** [comment]
Sure, I'd be happy to accept PRs for missing widgets in GTK and libadwaita. I would not add any widgets that are already deprecated though.

Also: 100th issue/discussion/PR :tada:

**Comment by PhilippMDoerner** [comment]
In that case first 2 tables for an overview over which Widgets yet to be wrapped.
That list already has the deprecated widgets stripped out of it.

### GTK
- [x] Progress Bar
- [ ] Scrollbar (in PR)
- [x] MediaControls
- [x] WindowControls
- [x] EmojiChooser
- [x] SearchEntry (Not easy to do. It requires passing it a GtkWidget, which currently is not possible. Only Partially done)
- [x] PasswordEntry (Not easy to do. It requires wrapping GMenu which to do it "right" requires tree diffing etc. etc.. Only Partially done)
- [x] EditableLabel
- [x] CenterBox
- [ ] SearchBar (Not easy to do, it requires passing it a reference to another GtkWidget which currently is not possible.)
- [x] ActionBar
- [ ] NoteBook (in PR)
- [ ] Stack (in PR)
- [ ] StackSwitcher
- [ ] StackSidebar
- [ ] PageSetupUnixDialog
- [ ] PrintUnixDialog
- [ ] ShortcutsWindow (in PR)

### Adw
- [x] Toast Overlay (in PR)
- [x] Banner (since 1.3)
- [x] SwitchRow (since 1.4)
- [x] PasswordEntryRow (since 1.2)
- [ ] SpinRow (since 1.4)
- [x] PreferencesPage (in PR)
- [ ] PreferencesWindow (in PR)
- [ ] NavigationView (since 1.4)
- [ ] NavigationSplitView (since 1.4)
- [x] OverlaySplitView (since 1.4) (in PR)
- [ ] Carousel (+ CarouselIndicatorDots + CarouselIndicatorLines)
- [ ] ViewSwitcher
- [ ] ViewSwitcherBar
- [ ] TabBar (+ TabOverview + TabButton)
- [ ] BreakpointBin (Since 1.4)
- [ ] ToolbarView (Since 1.4) (in PR)
- [x] HeaderBar
- [x] Window
- [x] ButtonContent
- [x] Bin

**Reply by can-lehmann** [reply]
Be sure to check the GTK4 docs for deprecations, e.g. https://docs.gtk.org/gtk4/class.Statusbar.html and https://docs.gtk.org/gtk4/class.InfoBar.html are both deprecated.

**Reply by PhilippMDoerner** [reply]
One goes over that list, the first couple classes that are deprecated don't show up on it, you make one assumption based on that, that GTK cleaned up the WidgetGallery and BAM - GTK docs hit you with the "We removed *some* deprecated widgets from the WidgetGallery, but not all of them!"

... I'll go over that dang list again

**Reply by PhilippMDoerner** [reply]
List updated >_> never trust an assumption you make about the GTK docs is what I learned from this.
Another wrong assumption one would make would be "One Heading, one Widget", which also is not correct as e.g. Carousel hides *3* Widgets inside of itself, Tabs as well.

**Reply by PhilippMDoerner** [reply]
Ohhhhh you call the adwaita Window a Windowsurface?
Any other Widgets that fall into the category of renamed widgets?

**Reply by can-lehmann** [reply]
I don't think so. `adw.Window` was renamed to `WindowSurface` for convenience, as otherwise there would be name conflicts with `owlkettle.Window`.

**Comment by PhilippMDoerner** [comment]
Memo for me:
Once the Widget List above has been worked through, skim through [the general docs page](https://docs.gtk.org/gtk4/) to look for features that might still need implementing.
Drag and Drop seems like a decent example for a non-obvious one.
